File: library/src/main/java/zhan/library/widget/BodyFrameLayout.java
package zhan.library.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.widget.FrameLayout;
import zhan.library.ScrollState;
import zhan.library.delegate.ScrollBodyDelegate;
import zhan.library.listener.OnStateChangeListener;


public class BodyFrameLayout extends FrameLayout {

    private ScrollBodyDelegate mScrollBodyDelegate;

    public BodyFrameLayout(Context context) {
        super(context);
        init();
    }

    public BodyFrameLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public BodyFrameLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        mScrollBodyDelegate = new ScrollBodyDelegate(this);
    }

    @Override
    public void computeScroll() {
        mScrollBodyDelegate.computeScroll();
    }

    public void setMaxOffset(int offset) {
        mScrollBodyDelegate.setMaxOffset(offset);
    }

    public ScrollState getState() {
        return mScrollBodyDelegate.getState();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return mScrollBodyDelegate.onTouchEvent(event);
    }

    public void hide(int time) {
        mScrollBodyDelegate.hide(time);
    }

    public void open() {
        mScrollBodyDelegate.open();
    }

    public void setOnStateChangeListener(OnStateChangeListener listener) {
        mScrollBodyDelegate.setOnStateChangeListener(listener);
    }
}

File: double-pull-delegate/src/main/java/zhan/pull/widget/BodyFrameLayout.kt
package zhan.pull.widget

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.FrameLayout
import zhan.pull.ScrollState
import zhan.pull.delegate.ScrollBodyDelegate
import zhan.pull.listener.OnStateChangeListener


class BodyFrameLayout : FrameLayout {

    private var mScrollBodyDelegate: ScrollBodyDelegate? = null

    val state: ScrollState
        get() = mScrollBodyDelegate!!.state

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        mScrollBodyDelegate = ScrollBodyDelegate(this)
    }

    override fun computeScroll() {
        mScrollBodyDelegate!!.computeScroll()
    }

    fun setMaxOffset(offset: Int) {
        mScrollBodyDelegate!!.setMaxOffset(offset)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return mScrollBodyDelegate!!.onTouchEvent(event)
    }

    fun hide(time: Int) {
        mScrollBodyDelegate!!.hide(time)
    }

    fun open() {
        mScrollBodyDelegate!!.open()
    }

    fun setOnStateChangeListener(listener: OnStateChangeListener) {
        mScrollBodyDelegate!!.setOnStateChangeListener(listener)
    }
}
